File: librarymanager/librarymanager-be/src/main/java/ru/elenacherev/librarymanager/domain/entity/BookEntity.java
package ru.elenacherev.librarymanager.domain.entity;

import lombok.*;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.HashSet;
import java.util.Set;

import static javax.persistence.GenerationType.IDENTITY;

@Entity
@Table(name = "BOOKS")
@Getter
@NoArgsConstructor
//@RequiredArgsConstructor
@Setter
public class BookEntity {
    @Id
    @Column(name = "BOOK_ID")
    @GeneratedValue(strategy = IDENTITY) //уникальный ID, генерируется сам в БД
    @NotNull
    @Setter(AccessLevel.NONE)
    private Long bookId;

    @Version
    @Column(name = "VERSION")
    private int version; //версия, инкрементится при редактировании записи

    @Column(name = "IS_RESERVED")
    private boolean isReserved;

    @Column(name = "IS_AVAILABLE")
    private boolean isAvailable;

    @Column(name = "SHELF_NUMBER")
    private int shelfNumber;

    @ManyToOne
    @JoinColumn(name = "EDITION_ID")
    private EditionEntity edition;

    @OneToMany(mappedBy = "book", cascade = CascadeType.ALL, orphanRemoval = true)
    @Setter(AccessLevel.NONE)
    private Set<BookUsingEntity> bookUsings = new HashSet<BookUsingEntity>();
}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/domain/entity/BookEntity.kt
package ru.elenacherev.librarymanager.domain.entity

import java.util.*
import javax.persistence.CascadeType
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table
import javax.persistence.Version
import javax.validation.constraints.NotNull

@Entity
@Table(name = "BOOKS")
data class BookEntity(

    @Id
    @Column(name = "BOOK_ID")
    @GeneratedValue
    @get:NotNull
    var bookId: Long? = null,

    //версия, инкрементится при редактировании записи
    @Version
    @Column(name = "VERSION")
    var version: Int = 0,

    @Column(name = "IS_RESERVED")
    var isReserved: Boolean = false,

    @Column(name = "IS_AVAILABLE")
    var isAvailable: Boolean = false,

    @Column(name = "SHELF_NUMBER")
    var shelfNumber: Int = 0,

    @ManyToOne
    @JoinColumn(name = "EDITION_ID", nullable = false)
    var edition: EditionEntity,

    @OneToMany(mappedBy = "book", cascade = [CascadeType.ALL], orphanRemoval = true)
    val bookUsings: MutableSet<BookUsingEntity> = mutableSetOf()
)